### DISCUSSION TITLE
Sitemaps with Incremental Static Regeneration

### DISCUSSION THREAD

**Discussion by MichaelBurgess** [vercel/next.js#24368] [Help]
How are people handling generation/updating of sitemaps when they use an ISR-based approach to next.js sites?

I use `next-sitemap`, but that relies on having all the static pages to generate. In the ISR model, all you know about is the new/updated page, so I'm curious how others are attacking this?

**Comment by gomflo** [comment]
`next-sitemap` has an option for generating dynamic/server-side sitemaps:

https://github.com/iamvishnusankar/next-sitemap#generating-dynamicserver-side-sitemaps

does it help?

**Reply by MichaelBurgess** [reply]
Thanks @gomflo - possibly that may work, although the overhead of me getting all the paths dynamically on the server may be pretty high and I'm worried about the time needed to serve up the sitemap. Ideally I'd want a stale while revalidate approach to this.

I run a build job that builds a cache for the next.js site to generate the static pages from, so it's just occurred to me that I should be generating the sitemap as part of that, then I can try and use an ISR-based solution to statically generate the sitemap in my site, which during revalidate pulls in the new version from my build cache, or I could just do server-generated sitemap that pulls from the build cache every time.

I *think* that will work, but proof of the pudding...

**Reply by BrunoQuaresma** [reply]
Hi, did you make this work? I see next-sitemap only works with Server Props not Static Props :/

**Reply by MichaelBurgess** [reply]
I didn't get any luck and haven't gone down the ISR route yet - so still unresolved for me.

**Reply by mutahhar-ahmad** [reply]
@MichaelBurgess Were you able to achieve this?

**Comment by rserafim** [comment]
Has anyone managed to figure out a way? My project generates over 5 million pages after build.

**Comment by franciscomntll** [comment]
Hola! Estoy presentando el mismo problema. El proyecto es un periodico digital, y maneja un alto volumen de creacion de paginas al dia. 
Tengo tiempos de hasta >30 segundos para la entrega del sitemap dinamico. Una locura, y de seguro el crawler de google no esperara eso. Por tanto he pensado en usar getStaticProps para crear un sitemap estatico y regenerar con el revalidate. Sin embargo no encuentro forma de que funcione.

Algun avance con sus proyectos?

**Reply by rserafim** [reply]
I researched that Google accepts up to 50,000 pages per xml. I separated my sitemaps, in my case by state, using getStaticPaths. I haven't deployed it yet. follow this thread https://github.com/iamvishnusankar/next-sitemap/issues/216

**Comment by zaarheed** [comment]
Is there a way to render XML in a Next.js page? I just quickly tried adding a `sitemap.xml.js` file under my `/pages` directory and it seems to load fine at `localhost:3000/sitemap.xml`. I haven't check this after deploying to vercel yet (wondering if there are any quirks).

Could it be possible to set this up as an ISR page with `refresh: 1` to serve the sitemap?

**Reply by Mrkk108** [reply]
Hello , did you find anything?

**Comment by bdDavin** [comment]
I found this https://dev.to/timrichter/dynamic-sitemap-with-next-js-41pe. Where he uses Cache Control header in getServerSideProps to make it like a getStaticProps with revalidate
